aospy/automate.py

```diff
@@ -0,0 +1,263 @@
+
+    From http://stackoverflow.com/a/26853961/1706640
+    """
+    result = {}
+    for dictionary in dict_args:
+        result.update(dictionary)
+    return result
+
+
+def _input_func_py2_py3():
+    """Find function for reading user input that works on Python 2 and 3.
+
+    See e.g. http://stackoverflow.com/questions/21731043
+    """
+    try:
+        input = raw_input
+    except NameError:
+        import builtins
+        input = builtins.input
+    return input
+
+
+class AospyException(Exception):
+    """Base exception class for the aospy package."""
+    pass
+
+
+def _user_verify(input_func=_input_func_py2_py3(),
+                 prompt='Perform these computations? [y/n] '):
+    """Prompt the user for verification."""
+    if not input_func(prompt).lower()[0] == 'y':
+        raise AospyException('Execution cancelled by user.')
+
+
+def _get_all_objs_of_type(type_, parent):
+    """Get all attributes of the given type from the given object.
+
+    Parameters
+    ----------
+    type_ : The desired type
+    parent : The object from which to get the attributes with type matching
+        'type_'
+
+    Returns
+    -------
+    A set (possibly empty) of attributes from 'parent'
+    """
+    return set([obj for obj in parent.__dict__.values()
+                if isinstance(obj, type_)])
+
+
+class CalcSuite(object):
+    """Suite of Calc objects generated from provided specifications."""
+
+    _CORE_SPEC_NAMES = {_OBJ_LIB_STR, _PROJECTS_STR, _MODELS_STR, _RUNS_STR}
+    _AUX_SPEC_NAMES = {_VARIABLES_STR,
+                       _REGIONS_STR,
+                       'date_ranges',
+                       'input_time_intervals',
+                       'input_time_datatypes',
+                       'input_time_offsets',
+                       'input_vertical_datatypes',
+                       'output_time_intervals',
+                       'output_time_regional_reductions',
+                       'output_vertical_reductions'}
+    _NAMES_SUITE_TO_CALC = {
+        _PROJECTS_STR: 'proj',
+        _MODELS_STR: 'model',
+        _RUNS_STR: 'run',
+        _VARIABLES_STR: 'var',
+        _REGIONS_STR: 'region',
+        'date_ranges': 'date_range',
+        'input_time_intervals': 'intvl_in',
+        'input_time_datatypes': 'dtype_in_time',
+        'input_time_offsets': 'time_offset',
+        'input_vertical_datatypes': 'dtype_in_vert',
+        'output_time_intervals': 'intvl_out',
+        'output_time_regional_reductions': 'dtype_out_time',
+        'output_vertical_reductions': 'dtype_out_vert',
+    }
+
+    def __init__(self, calc_suite_specs):
+        self._specs_in = calc_suite_specs
+        self._obj_lib = self._specs_in[_OBJ_LIB_STR]
+
+    def _get_requested_spec(self, obj, spec_name):
+        """Helper to translate user specifications to needed objects."""
+        requested = self._specs_in[spec_name]
+        if isinstance(requested, str):
+            return [_get_attr_by_tag(obj, requested, spec_name)]
```

Comment:
Since the aospy core variable attributes use ```utils.io.dict_name_keys```, this returns a list of a single dictionary, while the line below the ```else``` statement returns a list of aospy core objects (this inconsistency causes problems).  Should we do away with using ```dict_name_keys``` in the core objects themselves, and just assign lists to their ```models``` and ```runs``` attributes?